src/VisualStudio/Core/Def/Implementation/ChangeSignature/AddParameterTextViewModelProvider.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation.IntellisenseControls;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Projection;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ChangeSignature
+{
+    [Export(typeof(ITextViewModelProvider))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(VisualStudioChangeSignatureOptionsService.AddParameterTextViewRole)]
+    internal class AddParameterTextViewModelProvider : ITextViewModelProvider
+    {
+        [Import]
+        public IProjectionBufferFactoryService ProjectionBufferFactoryService { get; set; }
+
+        public ITextViewModel CreateTextViewModel(ITextDataModel dataModel, ITextViewRoleSet roles)
+        {
+            var projectionSnapshot = (IProjectionSnapshot)dataModel.DataBuffer.CurrentSnapshot;
+            // There are five spans: 
+            // 0. From the start of the document and to the inserted comma before the insertion.
+            // 1. The insertion span for the type - we need to catch it for AddParameterTypeTextViewRole.
+            // 2. the space span
+            // 3. The parameter name span.
+            // 4. The rest of the document.
+            // Please note that for VB we should use another structure: start, name, " AS ", type, rest
```

Comment:
Shouldn't this be doing the other thing for VB then?

More generally I'm wondering if we have some other options here to avoid having to mess with all the roles. You already have the function that takes the IProjectSnapshot and returns the list of tracking spans. If those instead returned a TrackingSpan plus optional marker of some kind, then instead of having to hardcode the constants here you could just keep the list of tracking spans and figure out SourceSpan here is involved. I can also think of a silly hack to instead then pass the marker name through the ITextViewRoleSet and you don't have to make any of this be specific to AddParameter at all.

(That last bit might require explanation in person...)

Reply:
We perform separation of C# and VB on other level. Just updated the comment.